File: rvc_service/RVCSDK/examples/CSharp/GetCorrespondMap.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using RVC_CSharp;

namespace RVC_CSharp
{
    class Program
    {

        static void Main()
        {
            #region Step 0 , Start init system.

            if (false == RVC_CSharp.System.Init())
            {
                Console.WriteLine("Failed to init system.");
                Console.ReadKey();
                return;
            }
            else
            {
                Console.WriteLine("Successfully init system.");
            }

            #endregion

            #region Step 1 , Start Find Devices

            Console.WriteLine("Step 1 , Start Find Devices.");

            List<Device> devices = RVC_CSharp.System.ListDevices(SystemListDeviceType.GigE);

            if (devices == null || devices.Count <= 0)
            {
                Console.WriteLine("Can not find device.\nPlease Check Connection.");

                Console.WriteLine("System close.");
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }

            Console.WriteLine($"Find {devices.Count} devices.");

            #endregion

            #region Step 2 , Choose Target Device

            Console.WriteLine("Step 2 , Choose Target Device.");

            Console.WriteLine("List All Devices.");
            DeviceInfo info = new DeviceInfo();
            for (int i = 0; i < devices.Count; i++)
            {
                var deviceTmp = devices[i];
                DeviceInfo infoTmp = new DeviceInfo();
                deviceTmp.GetDeviceInfo(ref infoTmp);
                bool canConnected = deviceTmp.CheckCanConnected();
                Console.WriteLine("\n");
                Console.WriteLine($"{i}-设备名称--[name] = {infoTmp.name}");
                Console.WriteLine($"{i}-设备序列号--[sn] = {infoTmp.sn}");
                Console.WriteLine($"{i}-出厂日期--[factroydate] = {infoTmp.factroydate}");
                Console.WriteLine($"{i}-端口类型--[type] = {infoTmp.type}");
                Console.WriteLine($"{i}-端口号--[port] = {infoTmp.port}【只针对网络相机】");
                if (canConnected)
                {
                    Console.WriteLine($"{i}-主板类型--[boardmodel] = {infoTmp.boardmodel}");
                    Console.WriteLine($"{i}-支持的相机类型--[cameraid] = {infoTmp.cameraid}");
                    Console.WriteLine($"{i}-是否支持单相机--[support_x1] = {infoTmp.support_x1}");
                    Console.WriteLine($"{i}-是否支持双相机--[support_x2] = {infoTmp.support_x2}");
                    Console.WriteLine($"{i}-支持的投影颜色--[support_color] = {infoTmp.support_color}");
                    Console.WriteLine($"{i}-工作距离-近--[workingdist_near_mm] = {infoTmp.workingdist_near_mm}");
                    Console.WriteLine($"{i}-工作距离-远--[workingdist_far_mm] = {infoTmp.workingdist_far_mm}");
                    Console.WriteLine($"{i}-固件版本--[firmware_version] = {infoTmp.firmware_version}");
                    Console.WriteLine($"{i}-支持的拍摄模式--[support_capture_mode] = {infoTmp.support_capture_mode}");
                    Console.WriteLine($"{i}-是否支持防护盖控制--[support_protective_cover] = {infoTmp.support_protective_cover}");
                }
                Console.WriteLine($"{i}-是否可以连接 = {canConnected}");
                Console.WriteLine("\n");
            }

            Console.WriteLine("Input the id of device that you want to connect.");

            Console.Write("ID : ");
            string str = Console.ReadLine();
            var parseRet = int.TryParse(str, out int id);
            if (!parseRet || id < 0 || id >= devices.Count)
            {
                Console.WriteLine("Input is not valid");
                Console.WriteLine("System close.");
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }

            RVC_CSharp.Device device = devices[id];
            device.GetDeviceInfo(ref info);
            #endregion

            #region Step 3 , Open Camera

            if (false == device.IsValid() || false == device.CheckCanConnected())
            {
                Console.WriteLine("Device Can not be Connected.\nPlease Check Connection.");
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }

            if (!info.support_x2 || (info.support_capture_mode & CaptureMode.CaptureMode_SwingLineScan) == 0)
            {
                Console.WriteLine("Device not support X2 or SwingLineScan!");
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }
            if(device.IsFirmwareMatch() == false)
            {
                Console.WriteLine("Device firmware mismatch. Please use RVCManager to upgrade the firmware");
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }
            X2 camera = RVC_CSharp.X2.Create(device);
            camera.Open();
            if (false == camera.IsValid() || false == camera.IsOpen())
            {
                Console.WriteLine($"Failed to open camera .Error: { System.GetLastErrorMessage() } ");
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }
            #endregion

            #region Step 4 , Capture

            X2.CaptureOptions options = X2.CaptureOptions.Default();
            camera.LoadCaptureOptionParameters(ref options);

            options.capture_mode = CaptureMode.CaptureMode_SwingLineScan;
            options.line_scanner_scan_time_ms = 2500;
            options.line_scanner_min_distance = 400;
            options.line_scanner_max_distance = 800;
            options.line_scanner_exposure_time_us = 300;
            options.correspond2d = false;

            bool ret = camera.Capture(options);

            if (false == ret)
            {
                Console.WriteLine($"Failed to Capture .Error: { System.GetLastErrorMessage() } ");
                camera.Close();
                camera.Destroy();
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }

            DepthMap depth = camera.GetDepthMap();
            PointMap pointMap = camera.GetPointMap();

            if (options.capture_mode == CaptureMode.CaptureMode_SwingLineScan && !options.correspond2d)
            {
                CorrespondMap correspond_map =camera.GetCorrespondMap();
                Size sz = correspond_map.GetSize();
                int select_point3d_index = (sz.height * sz.width) / 2;

                unsafe
                {
                    double* dataPtr = (double*)correspond_map.GetDataPtr();
                    double correspond2d_x = dataPtr[2 * select_point3d_index];
                    double correspond2d_y = dataPtr[2 * select_point3d_index + 1];
                    Console.WriteLine($"The {select_point3d_index}th 3D point corresponds to the 2D pixel: ({correspond2d_x}, {correspond2d_y})");
                }
            }
            else
            {
                Console.WriteLine("CorrespondMap is only effective when the mode is SwingLineScan mode and Correspond2d is set to false. In other cases, the 3D points already correspond to the 2D image.");
                camera.Close();
                camera.Destroy();
                RVC_CSharp.System.Shutdown();
                Console.ReadKey();
                return;
            }

            #endregion

            #region Step 5 , Data Process

            Directory.CreateDirectory($"./{info.name}-{info.sn}");


            // Depth
            string depthFile = $"./{info.name}-{info.sn}/Depth.tiff";
            Console.WriteLine($"Save Depth,path = {depthFile}");
            depth.SaveDepthMap(depthFile);

            // Point Cloud 
            string pointCloudFile = $"./{info.name}-{info.sn}/PointCloud.ply";
            Console.WriteLine($"Save PointMap,path = {pointCloudFile}");
            pointMap.SavePlyBinary(pointCloudFile);

            Console.WriteLine("Successfully Process Data .");

            #endregion

            #region Step 6 , Close And Release Device

            camera.Close();
            camera.Destroy();

            RVC_CSharp.System.Shutdown();
            Console.WriteLine("System close.");

            Console.ReadKey();
            #endregion

            return;
        }

    }
}

File: rvc_service/RVCSDK/examples/Python/GetCorrespondMap.py
import PyRVC as RVC
import numpy as np
import cv2
import os
from Utils.Tools import *

#  Note:CorrespondMap is only effective when the mode is SwingLineScan mode and correspond2d is set to false.
#  In other cases, the 3D points already correspond to the 2D image.

if __name__ == "__main__":
    # Initialize RVC system
    RVC.SystemInit()

    # Choose RVC Camera type (USB, GigE or All)
    opt = RVC.SystemListDeviceTypeEnum.All

    # Scan all RVC Camera devices
    ret, devices = RVC.SystemListDevices(opt)
    print("RVC Camera devices number:", len(devices))

    # Find whether any RVC Camera is connected or not
    if len(devices) == 0:
        print("Can not find any RVC Camera!")
        RVC.SystemShutdown()
        exit(1)    
    if devices[0].IsFirmwareMatch() == False:
        print("device firmware mismatch, Please use RVCManager to upgrade the firmware")
        RVC.SystemShutdown()
        exit(1)    
    # Create and open RVC Camera
    device = devices[0]
    x = RVC.X2.Create(device)
    x.Open()
    if x.IsOpen() == False:
        print("Failed to open camera! Please check whether the camera is connected and make sure it is not occupied and supports X2.")
        RVC.X2.Destroy(x)
        RVC.SystemShutdown()
        exit(1)

    ret, info = device.GetDeviceInfo()
    camera_id = RVC.CameraID_Left
    if info.support_extra:
        camera_id = RVC.CameraID_Extra
        
    # Print ExposureTime Range
    _, exp_range_min, exp_range_max = x.GetExposureTimeRange()
    print("ExposureTime Range:[{}, {}]".format(exp_range_min, exp_range_max))

    # Set capture parameters
    cap_opt = RVC.X2_CaptureOptions()
    ret,cap_opt = x.LoadCaptureOptionParameters()
    # Set capture mode
    cap_opt.capture_mode = RVC.CaptureMode_SwingLineScan
    # Set the scan time. The longer the scan time, the denser the point cloud will be.
    cap_opt.line_scanner_scan_time_ms = 2500
    # Set exposure time
    cap_opt.line_scanner_exposure_time_us = 300
    # Set minimum distance
    cap_opt.line_scanner_min_distance = 400
    # Set maximum distance
    cap_opt.line_scanner_max_distance = 800
    # Set whether to enable point clouds to correspond to 2D images.
    # If set to true, a depth map will be generated, otherwise no depth map will be generated.
    cap_opt.correspond2d = False

    # Capture a point map and a image.
    ret2 = x.Capture(cap_opt)

    # Create saving address of image and point map.
    save_dir = "Data"
    if not os.path.exists(save_dir):
        os.makedirs(save_dir)

    if ret2 == True:
        # Get image data. choose left or right side. the point map is map to left image.
        img = x.GetImage(camera_id)

        # Save image
        if img.SaveImage(save_dir + "/image.png"):
            print("Save image successed!")
        else:
            print("Save image failed!")

        # Save point map (m) to file.
        pm = x.GetPointMap()
        if pm.Save("{}/test.ply".format(save_dir), RVC.PointMapUnitEnum.Meter):
            print("Save point map successed!")
        else:
            print("Save point map failed!")
        
        if(cap_opt.capture_mode == RVC.CaptureMode_SwingLineScan and cap_opt.correspond2d == False):
            width = pm.GetSize().cols
            height = pm.GetSize().rows
            correspond_map = np.array(x.GetCorrespondMap(), copy=False)
            correspond_map = np.array(correspond_map, copy=False).reshape(-1, 2)
            select_point3d_index = (width * height) // 2
            correspond2d_x = correspond_map[select_point3d_index,0]
            correspond2d_y = correspond_map[select_point3d_index,1]
            print(
                f"The {select_point3d_index}th 3D point corresponds to the 2D pixel: ({correspond2d_x}, {correspond2d_y})")
        else:
            print("CorrespondMap is only effective when the mode is SwingLineScan mode and correspond2d is set to false.In other cases, the 3D points already correspond to the 2D image.")
        
        
    else:
        print("RVC Camera capture failed!")
        print(RVC.GetLastErrorMessage())
        x.Close()
        RVC.X2.Destroy(x)
        RVC.SystemShutdown()
        exit(0)
    # Close RVC Camera
    x.Close()

    # Destroy RVC Camera
    RVC.X2.Destroy(x)

    # Shut Down RVC System
    RVC.SystemShutdown()
